external_search.py

```diff
@@ -882,8 +577,748 @@ def create_new_version(cls, search_client, base_index_name, version=None):
+        genre, genre_match = KeywordBasedClassifier.genre_match(query_string)
+        if genre:
+            query_string = self.add_match_query(
+                genre.name, 'genres.name', query_string, genre_match
+            )
+
+        # Handle the 'young adult' part of 'young adult romance'
+        audience, audience_match = KeywordBasedClassifier.audience_match(
+            query_string
+        )
+        if audience:
+            query_string = self.add_match_query(
+                audience.replace(" ", ""), 'audience', query_string,
+                audience_match
+            )
+
+        # Handle the 'nonfiction' part of 'asteroids nonfiction'
+        fiction = None
+        if re.compile(r"\bnonfiction\b", re.IGNORECASE).search(query_string):
+            fiction = "Nonfiction"
+        elif re.compile(r"\bfiction\b", re.IGNORECASE).search(query_string):
+            fiction = "Fiction"
+        query_string = self.add_match_query(
+            fiction, 'fiction', query_string, fiction
+        )
+
+        # Handle the 'grade 5' part of 'grade 5 dogs'
+        age_from_grade, grade_match = GradeLevelClassifier.target_age_match(
+            query_string
+        )
+        if age_from_grade and age_from_grade[0] == None:
+            age_from_grade = None
+        query_string = self.add_target_age_query(
+            age_from_grade, query_string, grade_match
+        )
+
+        # Handle the 'age 10 and up' part of 'divorce age 10 and up'
+        age, age_match = AgeClassifier.target_age_match(query_string)
+        if age and age[0] == None:
+            age = None
+        query_string = self.add_target_age_query(age, query_string, age_match)
+
+        self.final_query_string = query_string.strip()
+
+        if len(query_string.strip()) == 0:
+            # Someone who searched for 'young adult romance' ended up
+            # with an empty query string -- they matched an audience
+            # and a genre, and now there's nothing else to match.
+            return
+
+        # Someone who searched for 'asteroids nonfiction' ended up
+        # with a query string of 'asteroids'. Their query string
+        # has a field match component and a query-type component.
+        #
+        # What is likely to be in this query-type component?
+        #
+        # In theory, it could be anything that would go into a
+        # regular query. So would be a really cool place to
+        # call Query() recursively.
+        #
+        # However, someone who does this kind of search is
+        # probably not looking for a specific book. They might be
+        # looking for an author (eg, 'science fiction iain
+        # banks'). But they're most likely searching for a _type_
+        # of book, which means a match against summary or subject
+        # ('asteroids') would be the most useful.
+        if (self.final_query_string
+            and self.final_query_string != self.original_query_string):
```

Comment:
it's possible these could be different because only the final string is stripped.